### PR TITLE
Cleaned constructor parameters for Indexer and ScalaCli

### PR DESCRIPTION
**Pull request by Abhijeetmishr**
replaced () =>in Indexer and ScalaCli classes.

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/Indexer.scala [hunk 1]
```diff
@@ -46,35 +46,35 @@ import org.eclipse.lsp4j.Position
...
+    symbolSearch: MetalsSymbolSearch,
+    buildTools: BuildTools,
+    formattingProvider: FormattingProvider,
     fileWatcher: FileWatcher,
-    focusedDocument: () => Option[AbsolutePath],
+    focusedDocument: Option[AbsolutePath],
     focusedDocumentBuildTarget: AtomicReference[b.BuildTargetIdentifier],
     buildTargetClasses: BuildTargetClasses,
-    userConfig: () => UserConfiguration,
+    userConfig: UserConfiguration,
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/Indexer.scala:77]: UserConfig should staty with () since it's a var and might change.

**Review by tgodzik — CHANGES_REQUESTED** [review]
Looks like the tests are failing, most likely some of the providers are not initialized properly. 

I would suggest chanhing one by one and running one of the failing tests locally to see at which point it break. You might need to change the order of the providers in that case.

File: metals/src/main/scala/scala/meta/internal/metals/Indexer.scala (hunk 1 above)
Comment (Abhijeetmishr) [reply metals/src/main/scala/scala/meta/internal/metals/Indexer.scala:77]: @tgodzik should I revert these and start with one by one ?
And how to perform tests for a single change ?

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/Indexer.scala:77]: If you run `sbt` you should be able to see examples on how to run a single test in the welcome view. You can also looks at the https://scalameta.org/metals/docs/contributors/getting-started site.

I would revert all of the changes and do it one by one or a couple at a time.

Sorry for not responding earlier, but I was off last week and only digging through emails right now.

**Comment by odisseus** [comment]
These arguments are passed as zero-argument functions (having type `() => T` instead of `T`), because their values may change many times during the object's lifetime. For example, `UserConfiguration` is passed on to the code that initializes and manages build servers. That code should see the configuration as it is at the time when the server needs to be spun up, rather than as it was when the Metals plugin was being initialized. 

Bluntly changing the argument types without handling that mutability was never going to work.

**Pull request closed without merging**